Refactor client_name handling in mlfi_connect

prehor opened the pull request:
Spamassassin's RDNS_NONE check requires in the synthesized Received: header Forward-confirmed reverse DNS (FCrDNS) client name or "unknown" if the FCrDNS is not available.

Sendmail sends the reverse DNS or IP address in square brackets both in the client_host argument and  the {client_name} macro. If the client name is FCrDNS, macro {client_reverse} is set to `OK`, otherwise to `FORGED`, `TEMP` or `FAIL`. 

Postfix sends a FCrDNS or "unknown" using the {client_name} macro.

Related-to: #1, #3

Signed-off-by: Petr Řehoř <rx@rx.cz>

prehor committed:
Refactor mlfi_connect client_name handling

Spamassassin's RDNS_NONE check requires in the synthesized Received: header verified reverse DNS hostname or "unknown" if the reverse DNS is not available.

Sendmail sends an IP address in square brackets in the client_host argument and in the {client_name} macro, if the verified reverse DNS is not available

Postfix sends a verified reverse DNS or "unknown" using the {client_name} macro.

Signed-off-by: Petr Řehoř <rx@rx.cz>

Bluewind commented on amavisd-milter/mlfi.c:
```diff
@@ -331,20 +338,23 @@ mlfi_connect(SMFICTX *ctx, char *hostname, _SOCK_ADDR * hostaddr)
+	client_name = "unknown";
+    }
+    logqidmsg(mlfi, LOG_DEBUG, "client name: %s", client_name);
+    if ((mlfi->mlfi_client_name = strdup(client_name)) == NULL) {
```
I guess you save the {client_name} macro's value here so that it can later be used for the Received header, but right now that still uses `mlfi_client_host`. Is that intentional or will you change the Received header code to use _name?

If the above is intentional: I'm not sure what values currently end up in the `_host` variable in the different cases and if they are want we want for spamassassin or not. I think that the old code worked fine for me when the hostname verification succeeded, so postfix probably sent the hostname via the parameter. When the verification failed it was somehow broken, but I don't remember how. I guess it might have sent NULL or an empty string, which means that your new code sets it to ''unknown". Do you know if that is what postfix does?

prehor replied:
Thanks, I forgot to use client_name in Received: header.

Bluewind replied:
Further down in the file there's a line where client_host is sent to amavis. It's around line 1191. Should this also be changed? I have no idea what that is used for and what the implication is.
```
amavisd_request(mlfi, "client_name", mlfi->mlfi_client_host)
```

prehor replied:
This is AM.PDP communication with the amavis daemon. Amavis uses it for logging and for some amavis checks. Spamassassins RDNS_NONE check uses the first trusted Received: header.

Bluewind reviewed (commented):
I don't really know much about the milter protocol nor about sendmail's/postfix' implementation of it, but I looked at your code anyways. My comments on it may or may not be useful. Feel free to not change anything if they seem wrong to you.

Apart from that I noticed that you have the "same" commit in this PR two times. Once with and once without the 'Signed-off-by' line. If you want to, you can clean this up with `git rebase -i master` and remove the "wrong" commit. If you've never used rebase before, be sure to read the documentation about rebase and maybe save the commit id of the current version in case you mess up the history. If you don't save the commit id, you can also get it via `git reflog`, but you might have to guess a bit since that's a long and potentially confusing list. With the id you can `git reset --hard $id` in case you broke the branch. Also make sure that you have commited all changes that are currently in your workdir before you rebase since you might lose some of them, especially when using `git reset --hard`.

prehor commented:
@Bluewind, thanks for your hint about git rebase. Now this PR looks better.

Could you please try that it works correctly with the postfix? I can only test it with a sendmail.

Bluewind commented on amavisd-milter/mlfi.c:
```diff
@@ -331,20 +338,23 @@ mlfi_connect(SMFICTX *ctx, char *hostname, _SOCK_ADDR * hostaddr)
-	    mlfi_setreply_tempfail(ctx);
-	    return SMFIS_TEMPFAIL;
-	}
+    if ((mlfi->mlfi_client_host = strdup(client_name)) == NULL) {
```
This should be `strdup(client_host)` instead of `strdup(client_name)`. _name is not defined yet.

prehor replied:
You are absolutely right.

prehor committed:
Fix typo

Signed-off-by: Petr Řehoř <rx@rx.cz>

Bluewind commented:
It appears to work correctly for me with postfix. Thanks!

prehor committed:
Invalidate client_name if client_resolve is not OK

Milter macro {client_resolve} contains "OK" if client_name contains Forward-confirmed reverse DNS (FCrDNS) name.

Signed-off-by: Petr Řehoř <rx@rx.cz>

prehor committed:
Send {client_resolve} macro at connect time

Signed-off-by: Petr Řehoř <rx@rx.cz>

prehor committed:
Improve comments

Signed-off-by: Petr Řehoř <rx@rx.cz>
